core/test/network-records-to-devtools-log.js

```diff
@@ -40,20 +46,181 @@ function headersArrayToHeadersDict(headersArray = []) {
+  }
+
+  // Start of request + relative timing must be <= responseReceivedTime and endTime.
+  const startTimes = [rendererStartTime, startTime, requestTime];
+  let maxStart = Math.max(...startTimes.filter(timeDefined));
+  maxStart = Number.isFinite(maxStart) ? maxStart : redirectResponseTimestamp; // Use redirectResponseTimestamp only as fallback.
+  if (timeDefined(maxStart)) {
+    if (timeDefined(responseReceivedTime)) {
+      assert(maxStart + relativeTimingMax <= responseReceivedTime, `request start (${maxStart}) plus relative 'timing' value (${relativeTimingMax}) exceeds 'responseReceivedTime' (${responseReceivedTime}) in test network record`); // eslint-disable-line max-len
+    }
+    if (timeDefined(endTime)) {
+      assert(maxStart + relativeTimingMax <= endTime, `request start (${maxStart}) plus relative 'timing' value (${relativeTimingMax}) exceeds 'endTime' (${endTime}) in test network record`); // eslint-disable-line max-len
+    }
+  }
+
+  // If all are defined, requestTime + receiveHeadersEnd === responseReceivedTime.
+  const {receiveHeadersEnd} = networkRecord.timing ?? {};
+  const start = timeDefined(startTime) ? startTime : requestTime;
+  if (timeDefined(start) && timeDefined(receiveHeadersEnd) && timeDefined(responseReceivedTime)) {
+    assert.equal(start + receiveHeadersEnd, responseReceivedTime, `request start (${start}) plus 'receiveHeadersEnd' (${receiveHeadersEnd}) does not equal 'responseReceivedTime' (${responseReceivedTime}) in test network record`); // eslint-disable-line max-len
+  }
+
+  return {
+    redirectResponseTimestamp,
+    rendererStartTime,
+    startTime,
+    requestTime,
+    receiveHeadersEnd,
+    responseReceivedTime,
+    endTime,
+    relativeTimingMax,
+  };
+}
+
+/**
+ * Takes the partial network timing and fills in the missing values so at least
+ * time is linear (if not realistic). The main timing properties on
+ * `NetworkRequest` and `requestTime` and `receiveHeadersEnd` on
+ * `NetworkRequest['timing']` will be computed. The other `timing` properties
+ * will not be automatically provided, but will be copied if provided, so they
+ * are up to the caller to order correctly. If no absolute times are provided,
+ * starts at `defaultStart`, finishes receiving headers `defaultTimingOffset` ms
+ * later, then endTime is `defaultTimingOffset` ms after that.
+ * Throws an error if conditions appear impossible to satisfy.
+ * @param {Partial<NetworkRequest>} values
+ * @return {NormalizedRequestTime}
+ */
+function getNormalizedRequestTiming(networkRecord) {
+  const extractedTimes = extractPartialTiming(networkRecord);
+
+  const possibleStarts = [
+    extractedTimes.startTime,
+    extractedTimes.requestTime,
+    extractedTimes.rendererStartTime,
+    // Note: since redirectResponseTimestamp (aka the redirect's endTime) is
+    // used as a last resort, some redirected requests may start before the
+    // redirect ends. Up to the caller to override if this matters.
```

Comment:
this isn't possible to fix without figuring out timing for all requests at once instead of individually, so they extra complexity didn't seem worth it.